### ISSUE TITLE
Bug: [unified-signatures] Recommends merging overloads that would then allow incorrect call signatures

### ISSUE THREAD

**Issue by requinix** [typescript-eslint/typescript-eslint#10962]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=PTAEAZQQwJwcwK4FsCmA7ALgZwLACgAzBNAYwwEsB7NUAyygCgEoAuUAN0vIBMBufEKACMAWgBM0eMnTYANKAwALFLXIwsGUOSygYKAI4I1KbviKkK1WvQZQ2GmOTRx5AIwD89jI%2BesOXPnwzYjIqGjpGAH0oT1AHJxdQSI8vHzg-Th5QAG9QAF8gvEElFQI1DVAGcgJoNABPJlAkBArXFW4UMrQTMxtiDq6TJn5CPrQBpxN5fs7J7mHCwRFlxbBwcUlEVEwseSh60BJKJFcnKEsaShr451ruUBnB00IQi9BXWFtYm8SUuO8EhkAiNghYwu9PtFvgDnPJktC0kCsrkCngBGA6igoIo4gg9AplKp1JoSPt3ipHnN5EpzgByHSUADWUDq%2BA%2BMAYlO68xG7M541m3OmAqeCzwQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdFADuY6E0jgwAXxBqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function foo(): void;
function foo(a: string, b?: string): void;
//                      ^^^^^^^^^^
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/unified-signatures": "warn",
  },
};
```

### tsconfig

```jsonc

```

### Expected Result

No lint message.

### Actual Result

Lint says the two overloads should be combined.

### Additional Info

tldr 1: yes, I'm talking about `undefined`, but (a) there's more to it than simply "NAB because optional `==` undefined", (b) the edits the lint rule wants me to make will cause code to allow something that wasn't allowed previously, and (c) the rule is reporting on a parameter that it doesn't make sense to report on.

tldr 2: some commentary in the Playground link

---

These two overloads
```typescript
function foo(): void;
function foo(a: string, b?: string): void;
```
are supposed to be combined into
```typescript
function foo(a?: string, b?: string): void;
```

The first problem with doing so is that `foo(undefined)` becomes a valid call signature when it wasn't before. Now yes, I know that a developer should understand that calling foo like that is almost* the same as just `foo()`, so let's ignore that one. (It's also been brought up in a past report or two about this lint rule.)

The second problem is that `foo(undefined, "bar")` becomes valid as well. In that case the whole "basically the same" argument backfires because a developer would probably expect this call to be meaningful - the `"bar"` is meaningful - when the implementation for `foo` is likely* something simple like `if (a === undefined) { ... }` and so completely ignoring the value.

(A workaround I've used in other cases has been to enable `ignoreDifferentlyNamedParameters`, which was really nice because it helped to point out that the meaning of the parameters was actually slightly different between the overloads, but obviously that doesn't work here.)

The main reason I would call this a bug and not simply "lol Javascript" is that Typescript will quite (un)happily complain about calling `foo(undefined, string)` when using the two overloads but will not complain about it when using the combined one. And I'd think one of the main principles of linting is that following its suggestions/fixes should not make invalid code become valid, right?

There's also a second reason this seems to be a bug: it's triggering on the optional `b?` parameter and I can't imagine why it should. There's [a lot of code](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/unified-signatures.ts) to sift through, but I'd speculate that it's detecting "overload with N required args / overload with N+1 required args" and triggering on the last parameter in the signature, when I think it should be accounting for the presence of optional parameters before deciding they can be combined.

I think my point's been made, but here's an interesting thing. Consider this:
```typescript
function foo(): void;
function foo(a: string): void;
function foo(a: string, b: string): void;
```
unified-signatures will say 1+2 can combine and 2+3 can combine, so naturally you'd think 1+2+3 could combine too. And currently, if you did merge a pair then the lint rule would agree and say to merge once more. But if that's incorrect and lint rule is fixed, if you merge one pair then you _won't_ get a lint message to merge again. Which could be surprising.

Except it _was_ incorrect. You can safely combine 1+2 or 2+3 but you _can't_ combine 1+2+3. For the reasons I gave earlier, which I'll repeat here as:

```typescript
// 1+2, 3
function foo(a?: string): void;
function foo(a: string, b: string): void; // foo(undefined, string) not okay

// 1, 2+3
function foo(): void;
function foo(a: string, b?: string): void; // foo(undefined, string) not okay

// 1+2+3
function foo(a?: string, b?: string): void; // foo(undefined, string) suddenly okay
```

---

\* Additional backstory that may be more relevant than I think: for better or worse, I tend to code the implementation of such overloaded functions like
```typescript
function foo(...args:
    | []
    | [a: string, b?: string]
): void {
    // branch on args.length
```

Merging the overloads then breaks their compatibility with the implementation, not just with `args` but with the code - now I have to support a `args.length == 1 && args[0] == undefined` case that didn't exist before. Yes, I know, "but what if someone calls the function incorrectly", but to that I say "[sorry, not sorry](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out)".

**Comment by kirkwaiblinger** [comment]
+1, this seems like a bug. The second call signature requires a non-undefined first parameter and the suggested combination doesn't reflect this.

**Comment by bradzacher** [comment]
My 2c: there isn't a bug here.

The rule is a stylistic rule that suggests that based on the types you could have a single signature and get the same overall signature.

In some codebases the merged signature might be preferred because it would simplify calling code.
EG with your split signatures I **_CANNOT_** call the function with an `undefined` first argument and instead would need to branch:

```ts
// your signatures force me to do:
if (arg1 != null) {
  foo(arg1, arg2);
} else {
  foo();
}

// the unified signatures would allow me to avoid this:
foo(arg1, arg2);
```

Stylistically you want to separate the two cases entirely.
And that's a totally fine choice to make!

The rule doesn't have a fixer here -- so it's just informing you that you could do this.
It's up to you to action the report or suppress the report with a disable comment.

**Comment by requinix** [comment]
Disclaimer: I'm verbose because that's just who I am, I promise I'm not trying to railroad any conversations with my walls of text 😅

@bradzacher 

> The rule is a stylistic rule that suggests that based on the types you could have a single signature and get the same overall signature.

> In some codebases the merged signature might be preferred because it would simplify calling code.

But that sounds like you _agree_ this is a bug:

1. A stylistic rule is something like tabs vs. spaces, or bracket placement, or semicolons (98% style, 2% not), and what makes them stylistic is the fact that accepting their fixes won't change the meaning of code.
2. Accepting unified-signatures's recommendation to merge the signatures means that the meaning of `foo(arg1, arg2)` has changed: it was not allowed before (for whatever reason) and it became allowed after.

I too agree that unified-signatures is a stylistic rule. Which is why its recommendation to merge these two particular signatures, resulting in a change in calling semantics for `foo`, is a bug.

> The rule doesn't have a fixer here -- so it's just informing you that you could do this.

Whether it has an actual fixer or not is an implementation detail; I wouldn't expect it to because merging code can be Really Hard, so even if it did then I wouldn't trust it and would rather merge the signatures manually.
(when I say "accept a fix", I mean accepting the change the rule is trying to make me do, not necessarily clicking a literal "Fix" button or running with `--fix` and getting the change made for me automatically)

While it doesn't have a fixer, the rule clearly has an intended outcome in mind: the two signatures get combined and a particular parameter (it highlights `b?` but obviously means `a` - I mentioned this in the bug report too) gets made optional.
The reason it attempts to do so is valid, and I respect and agree with its reason - thus why I don't simply turn the rule off - except _in this particular case_ I believe it's applying its reasoning incorrectly and coming up with a false positive.

> It's up to you to action the report or suppress the report with a disable comment.

I _hate_ `// eslint-disable`-ing lint rules for code, and I'm proud to say I've never found a case where I needed to do so. There's always another option more favorable to me than that.

But like I said, I agree with the rule in principle, I just think that it's making a mistake this time. A precisely-defined and _correctable_ mistake. So I'm "actioning" it, so to speak, by creating a bug report 😛

---

Regarding the example. what happens when I substitute `foo` with something that more closely resembles the real-world code that inspired this bug report? (it actually uses the `...args` trick I showed earlier, but I'll rewrite it to the more traditional style)
[Playground link](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgCpQig3gWAFDLIZwAmA9iADYCeyA2gNYTUBcyAzmFKAOYC6bdJmQAfDl14BufAF98%2BBJTjt2yACIQEwdsApDs%2BQgGIumQRgjS8h5AgqcoAVwRgyUABSmI5zAEpkuASEyGAAFtoAdCYWyAC8IRZWhHLWQRggJNDuvmwAbmTAJElEEBlZXj4QADTIoJngcGC6IAD8bA68Ocj5hVY26ZkeXm1oFjV1pWCNzSMdIDxdPSQBNoTAMMjuAIRe-oHBwWGRA%2BXh7FG7xQcYYI5QIFfIKau1ZQ1NFMgtLfEARL%2BPGBuTZ2ECcehMag1XJwSiOCB8ZBkDYAeQARgArTRgCKTbgQdieCy%2BPYvNYbTzUAAOEGR3Vh8LifzmPF%2BpKCB0IoPYZEoEAilDIPHcAAMACRYCbvZoyCWQmRxZASmFwiAyEW%2BR7JZAQSjsAwcznc3n8wXC8WSt5TD4gWVYeWKrAarWHM4RE4eFXwmoWqXWmWEZ1k4LGvkCoWiiV%2B6YUGTqzXBlIHJNPWTyPDcsAJYTxEAQADu6k02ma%2Bnc%2BwSjhADAA%2BgBGNgVwhoqBwEAIUL1%2BuN4N8uAwLt17vIX59mC-KovGSTw0ttsdrsAJh7hsIY6Xw9HEH745nnOQ66Hi5ry5HY5gu6ne6e164Vdrp6byDn7c7x4bK1XB%2B3A-fm-PF4TleNjTrICZ4F47qlIM2RWEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdFADuY6E0jgwAXxBqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

```typescript
  render(): void;
  render(tree: Tree, indentation?: string): void;

  render(tree?: Tree, indentation?: string): void {
    if (!tree) {
      this.render(this.#tree);
      return;
    }

    indentation ??= "";
    for (const [key, value] of Object.entries(tree)) {
      // render tree with recursive calls to render(child node, increased indentation)
```
Calling `render(maybeUndefinedVar, "string")` is not allowed by the separate signature declarations. That's a good thing because if I tried to do so anyway, my "string" indentation would get lost.
But the lint rule would have the two signatures merged to `render(tree?: Tree, indentation?: string)`, _not require any other code changes as a result_*, and end up allowing the function call, creating a sort of weird, _hidden_, buggy behavior where sometimes the rendered tree has the custom indentation and sometimes it does not.

---

\* Count that as a benefit of the `...args` syntax: changing the overload definitions also requires updating the implementation's parameter list, which then likely requires updating some amount of the function's code to match. Leaves no room for accidentally accepting a suggestion from unified-signatures that inadvertently affects the meaning of the function 👍

**Comment by JoshuaKGoldberg** [comment]
> verbose

I'm having a hard time finding the time to reading through the >1200 words already in this discussion. @requinix could you please shrink your points down to, say, 3-4 sentences? What's the tl;dr? (because, apologies, tl;dr)

**Comment by bradzacher** [comment]
> But that sounds like you agree this is a bug
> ...
> I too agree that unified-signatures is a stylistic rule. Which is why its recommendation to merge these two particular signatures, resulting in a change in calling semantics for foo, is a bug.

I do not agree. I was saying that you have an opinion of one style and the rule is recommending a different style. Style can mean more things than just pure non-runtime-impacting changes. As an example -- some codebases have the (sometimes linted) style of "follow the path of least indentation for blocks". That involves people physically writing code in a specific way so that they can de-nest `if`'s etc to reduce maximum indentation. That is code style, despite it changing the way code is written and executed.

Do not confuse "code formatting" with "code style". Code formatting (eg prettier) makes changes to code that do not change runtime semantics ever. Code style is a superset that historically includes code formatting, but also includes the way code is written.

---

> Whether it has an actual fixer or not is an implementation detail

Having a fixer is actually NOT an implementation detail.

There are many cases that this rule could autofix -- for example it's a very easy mechanical change to merge the signatures in your OP.

However autofixers have one golden rule: DO NOT BREAK THE BUILD. This means don't make a change that can introduce runtime breakages and typecheck errors. In this instance there would be no way for us to ensure that a signature change wouldn't break the build somewhere.
For example [if you're using `Parameters<T>` on the function signature then merging the signatures will change its returned type](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALngDcctgBuAKFElgRXWzwOKgoGcMMLKgDm5KjXoMGGAJ4AHBABUAjPAC88AAqwoAWxAYQMPgB45inIm4A+Rk3DQ4SNJlz5EhIlAD8-QcJiFNS09hbKAEwa2roGRibmCiBWtoxAA) and this change could break the build.

The absence of a fixer means that a human must make the conscious decision to make the change.
This means that the human can also make the decision to not make the change if they don't think that the merging achieves the desired outcome.

---

> I hate // eslint-disable-ing lint rules for code, and I'm proud to say I've never found a case where I needed to do so.

For sure -- be proud that you strive to write code that adheres to your linter config, but _never using disable comments ever_ isn't necessarily good.
But disable comments are a tool -- lint rules aren't always 100% correct and using disable comments is completely fine and valid.

For example in our codebase we have [over 170 files with disable comments](https://github.com/search?q=repo%3Atypescript-eslint%2Ftypescript-eslint%20%2Feslint-disable%2F&type=code) and each one of those is an case where we've intentionally violated a lint rule.

**Comment by requinix** [comment]
(sorry if there was something important to reply to and I didn't, I'm really hungry)

> As an example -- some codebases have the (sometimes linted) style of "follow the path of least indentation for blocks". That involves people physically writing code in a specific way so that they can de-nest `if`'s etc to reduce maximum indentation. That is code style, despite it changing the way code is written and executed.

As a devout follower of that exact principle, it is most definitely a matter of code style. It's one of the things that belong in disappointingly heavy-handed "Style Guide" documents - along with the standard formatting rules we all know and hate.
But any such refactoring under the excuse of "zomg reduce indentation!!1!" should have _zero_ effect on what the code truly means: the same set of inputs will result in the same actions performed in the same order, with the difference being that it prefers using things like `return` and `break` to affect _control flow_ instead of `if/else` or `switch`. It changes the way code is written but doesn't, in fact, change execution (at least not in a way that matters).

And that's exactly why it's a matter of code "style": the code has changed in a _literal_ way, perhaps even enough so that someone reading it might have to make adjustments to their understanding of how it works, but the code hasn't changed in a _fundamental_ way.

A lint rule that tries to enforce reduced indentation will come in one of two varieties: (a) it establishes a delineation at some arbitrary level of indentation N such that indentation <=N is "okay" but indentation >N is "not okay", or (b) it syntactically reads the code for known patterns that allow for refactoring.
The former is what you get when you want a simple and capricious solution, which rapidly turns into needing configuration because if you put 3 projects into a room then you'll get 4 decisions on how much indentation is okay.
The latter is better because it targets the actual situation that the underlying style is meant for, but it's also harder to implement. Theoretically, it could detect a function of `if { some code} else { lots of code }` and suggest refactoring to `if { some code; return } lots of code`. And it could do so safely. Which means it could, with even more effort, include a fixer.

unified-signatures is a style rule because following the suggestion (fixer or not) results in a _literal_ change to the code, perhaps even enough so that someone reading adjusts their understanding of it, but not a _fundamental_ change.
...is how it should be. except in the situation I've pointed out, the suggestion results in a fundamental change: `foo(undefined)` was not valid before but becomes valid after. It had no meaning before but is given a meaning after. And that means the lint rule has breached its authority over code _style_ and touched on matters of code _design_.

> `Parameters<T>`

I think you meant to have two different sets of functions there, so [here's an edited version](https://www.typescriptlang.org/play/?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALngDcctgBuAKFElgRXWzwOKgoGcMMLKgDm5KjXoMGGAJ4AHBABUAjPAC88AAqwoAWxAYQMPgB45inIm4A+Rk3DQ4SNJlz4ARrCJQA-P0FhMQpqWnsLZQAmDW1dAyMTcwUQK3gvGDsGIA) where T1 is reported as `[a: string]` and T2 is reported as `[a?: string | undefined]`.
The behavior for T1 doesn't make sense to me: the parameters for the first function are either `()` or `(string)` so the tuple should be `[] | [a: string]` - which would make it effectively (if you also include the established opinion that "optional `==` undefined") the same as T2.

IMO Parameters isn't well suited for overloaded functions: it reports the arguments according to the Javascript single-function interpretation of its arguments, rather than using a unioned set of tuples according to Typescript's understanding of the function. Problem is that the whole point of having overloaded function signatures is for situations where the single-function interpretation isn't good enough to understand a function's usage.

> For sure -- be proud that you strive to write code that adheres to your linter config, but never using disable comments ever isn't necessarily good.
> But disable comments are a tool -- lint rules aren't always 100% correct and using disable comments is completely fine and valid.

And that's fine. Normal, even. Those tools exist for specific reasons and pretending they don't exist doesn't help anybody.
I mean, my personal preferences happen to have a distaste for coffee, but I'm not going to advocate for people to switch to decaf. That said, if we're going to meet at a cafe for something, let's pick a place that also serves a good hot chocolate.

_Anyway_,

> A human should be reviewing the lint report and understanding the impact so that they can make all of the required changes implied by the signature change.

Correct. And when I first found this issue, that's what I did: review what it said, understand the impact, and decide whether to make the change. As one should always try to do when a linter reports something non-trivial - again, the tool exists for a specific reason.

The problem is that I reviewed what it said and decided that it didn't make sense _this particular time_. Compare that to another time when unified-signatures said I should combine two, and I realized that what I wanted was to enable the `ignoreDifferentlyNamedParameters` option and rename a parameter - not that the lint rule was wrong, but rather that there was a philosophical "bug" of a parameter with an inaccurate name, Or to another time (surely more than one) when I've had some overloaded function signatures, did some refactoring, tripped into triggering unified-signatures, and realized that a combined signature was correct and I just hadn't noticed it myself because I'd been too distracted with rearranging code.

Also, not to repeat this too much, _the lint highlights the wrong parameter and that alone suggests a logic bug somewhere_:
* If the code is correct then it will highlight the correct parameter,
* It is not highlighting the correct parameter,
* QED the code is not correct.

But this bug report isn't about the philosophy behind code linters, nor about the nuanced differences between what constitutes formatting vs. style, nor about coffee. This bug report is theoretically very simple: given the scenario of the two overloads using required and also optional parameters, is unified-signatures correct in its behavior of reporting that they (quote) "can be combined into one signature with an optional parameter"? (and if it is then secondary bug: it's highlighting the wrong parameter)

**Comment by JoshuaKGoldberg** [comment]
It's been 6 months without any activity on this issue. Which indicates either or both of:

* This edge case is rather rare and perhaps not worth us taking limited bandwidth away from other, higher-impact issues
* Reading the _multiple thousands of words_ already written in the issue is too daunting a task for most contributors

If someone has a **very minimal reproduction** and **3-4 sentence explanation** of it, please post a new issue with our template. We'd be happy to take a look. Cheers!
